File: coding/cses/RoundTrip.py
# https://cses.fi/problemset/task/1669/
def dfs(node, parent, visited, adj):
    st = [(node, -1)]
    parent[node] = -1
    while st:
        u, p = st.pop()
        if visited[u]:
            continue
        visited[u] = True
        for v in adj[u]:
            if v == p: continue
            if not visited[v]:
                parent[v] = u
                st.append([v,u])
            else:
                # print("cycle found")
                # construct cycle
                cycle = [v]
                cur = u
                while cur!=v and cur!=-1:
                    cycle.append(cur)
                    cur = parent[cur]
                # print(cycle)
                if cur == v and len(cycle)>=3:
                    cycle.reverse()
                    cycle.append(cycle[0]) # closed cycle
                    return cycle
                    
    return []
    

def main():
    n, m = map(int, input().split())
    adj = [[] for _ in range(n+1)]
    
    for i in range(m):
        u, v = map(int, input().split())
        adj[u].append(v)
        adj[v].append(u)
    
    vis = [False] * (n+1)
    parent = [-1]*(n+1)
    found = False
    for u in range(1,n+1):       
        cycle = dfs(u, parent, vis, adj)
        if len(cycle)>0:
            # print(cycle)
            found = True
            break
    if not found:
        print("IMPOSSIBLE")
    else:
        print(len(cycle))
        for x in cycle:
            print(x, end=" ")
        print()
        

if __name__ == "__main__":
    main()

File: coding/cses/round_trip.cpp
#include <vector>
#include <algorithm>
#include <iostream>
using namespace std;

vector<vector<int>> adj;
vector<int> parent;
int cycle_start, cycle_end;

bool dfs(int x, int p, vector<bool>& vis){
	vis[x] = true;
	for(int y:adj[x]){
		if(y==p) continue; //do not go to parent
		parent[y] = x;
		if(vis[y]){
			cycle_start = y;
			cycle_end = x;
			return true;
		}
		else{
			if(dfs(y, x, vis))
				return true;
		}
	}
	return false;
}


int main(){
	int n, m;
	cin >> n >> m;
	adj.resize(n+1);
	parent.resize(n+1);

	for(int i=0; i<m; i++){
		int a,b;
		cin>>a>>b;
		adj[a].push_back(b);
		adj[b].push_back(a);
	}

	cycle_start = -1;
	vector<bool> visited(n+1, false);
	for(int i=1; i<=n; i++){
		if(visited[i] == false)
			if(dfs(i, -1, visited))
				break;
	}

	if(cycle_start == -1){
		cout<<"IMPOSSIBLE\n";
		return 0;
	}

	vector<int> cycle;
	cycle.push_back(cycle_start);
	for(int v=cycle_end; v!=cycle_start; v = parent[v])
		cycle.push_back(v);
	cycle.push_back(cycle_start);
	// reverse(cycle.begin(), cycle_end());
	cout<<cycle.size()<<'\n';
	for(int x:cycle)
		cout<<x<<" ";
	cout << endl;

	return 0;
}